### PR TITLE
2115: More flexible merge PR review configuration

### PR DESCRIPTION
**Pull request by edvbld**
Hi all,

please review this patch that enables a bit more flexible configuration for checking merge pull requests. Today we can configure that a merge pull requests always should be checked and if that isn't configured then the `.jcheck/conf` from the repository is used. If we ever want to set `merge=check` in the `.jcheck/conf` file in repos then we will need a way to _disable_ running jcheck on merge pull requests (for example for projects using merge pull requests to sync in commits).

This patch makes it possible to configure that merge pull requests should either always be checked, never be checked or checked according to `.jcheck/conf` in the repo.

Thanks,
Erik

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace

### Issue
 * [SKARA-2115](https://bugs.openjdk.org/browse/SKARA-2115): More flexible merge PR review configuration (**Enhancement** - P4)

### Reviewers
 * [Erik Joelsson](https://openjdk.org/census#erikj) (@erikj79 - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/skara.git pull/1592/head:pull/1592` \
`$ git checkout pull/1592`

Update a local copy of the PR: \
`$ git checkout pull/1592` \
`$ git pull https://git.openjdk.org/skara.git pull/1592/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 1592`

View PR using the GUI difftool: \
`$ git pr show -t 1592`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/skara/pull/1592.diff">https://git.openjdk.org/skara/pull/1592.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/skara/pull/1592#issuecomment-1842614457)

### CODE REVIEW

File: bots/pr/src/main/java/org/openjdk/skara/bots/pr/MergePullRequestReviewConfiguration.java [hunk 1]
```diff
@@ -0,0 +1,29 @@
+/*
+ * Copyright (c) 2019, 2023, Oracle and/or its affiliates. All rights reserved.
```
Comment (zhaosongzs) [inline bots/pr/src/main/java/org/openjdk/skara/bots/pr/MergePullRequestReviewConfiguration.java:2]: Remove 2019?

**Review by zhaosongzs — COMMENTED** [review]
It's a reasonable change for me and it looks quite good to me!

File: bots/pr/src/main/java/org/openjdk/skara/bots/pr/MergePullRequestReviewConfiguration.java (hunk 1 above)
Comment (erikj79) [reply bots/pr/src/main/java/org/openjdk/skara/bots/pr/MergePullRequestReviewConfiguration.java:2]: No, the copyright lines from Oracle are either on the form `firstYear,` or `firstYear, lastYear,`. Never remove the first year.

File: bots/pr/src/main/java/org/openjdk/skara/bots/pr/PullRequestBot.java [hunk 2]
```diff
@@ -367,7 +367,7 @@ public String mlbridgeBotName() {
         return mlbridgeBotName;
     }
 
-    public boolean reviewMerge(){
+    public MergePullRequestReviewConfiguration reviewMerge(){
```
Comment (erikj79) [inline bots/pr/src/main/java/org/openjdk/skara/bots/pr/PullRequestBot.java:370]:
```suggestion
    public MergePullRequestReviewConfiguration reviewMerge() {
```

File: bots/pr/src/main/java/org/openjdk/skara/bots/pr/PullRequestBotFactory.java [hunk 3]
```diff
@@ -205,7 +205,16 @@ public List<Bot> create(BotConfiguration configuration) {
...
             if (repo.value().contains("reviewMerge")) {
-                botBuilder.reviewMerge(repo.value().get("reviewMerge").asBoolean());
+                MergePullRequestReviewConfiguration result = MergePullRequestReviewConfiguration.JCHECK;
+
+                var val = repo.value().get("reviewMerge").asString().toLowerCase().trim();
+                if (val.equals("always")) {
+                    result = MergePullRequestReviewConfiguration.ALWAYS;
+                } else if (val.equals("never")) {
+                    result = MergePullRequestReviewConfiguration.NEVER;
+                }
```
Comment (erikj79) [inline bots/pr/src/main/java/org/openjdk/skara/bots/pr/PullRequestBotFactory.java:215]: We should throw on an invalid configuration.

File: bots/pr/src/main/java/org/openjdk/skara/bots/pr/MergePullRequestReviewConfiguration.java (hunk 1 above)
Comment (edvbld) [reply bots/pr/src/main/java/org/openjdk/skara/bots/pr/MergePullRequestReviewConfiguration.java]: Thanks, fixed in latest commit!

File: bots/pr/src/main/java/org/openjdk/skara/bots/pr/PullRequestBot.java (hunk 2 above)
Comment (edvbld) [reply bots/pr/src/main/java/org/openjdk/skara/bots/pr/PullRequestBot.java]: Thanks, fixed in `HEAD`

File: bots/pr/src/main/java/org/openjdk/skara/bots/pr/PullRequestBotFactory.java (hunk 3 above)
Comment (edvbld) [reply bots/pr/src/main/java/org/openjdk/skara/bots/pr/PullRequestBotFactory.java:221]: Agree, fixed in latest commit.

**Review by erikj79 — APPROVED** [review]

**Comment by edvbld** [comment]
Thanks @zhaosongzs and @erikj79 for reviewing!

/integrate

**Pull request closed without merging**